che-incubator/chectl
# feat: Implement login manager and login/logout commands

Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### What does this PR do?
Before this change users had to provide Che server API URL and access token for each run of all commands that require authentication. This is not a user-friendly way to go. In this PR login manager is implemented and new commands introduced:
- `auth:login` logins user into a Che server. So, after login user can just run commands without Che API URL and access token parameters.
- `auth:logout` logs out user from current Che server
- `auth:get` - show current login info
- `auth:list` - shows all available logins
- `auth:use` - switches between login sessions
- `auth:delete` - deletes specified login session or all logins for the given Che server

Also, comfortable interface for using in commands is created.

The implementation supports 3 login methods via:
 - refresh token
 - username and password
 - oc login

Note, login manager does not (and should not) depend on `kubectl` or `oc` login information. It might use it, thought, to simplify login process.

Under the hood, login manager saves users refresh tokens. Also it is capable to automatically remove outdated ones.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16415

### How to test this PR?
At least two running instances of Eclispe Che is needed (`s1`, `s2`) with two users (`u1(admin user)`, `u2`) in each. One instance should have OAuth enabled on Openshift, where the second one should use password login on Kubernetes. For testing purposes I used `workspace:list` command, so it is helpful when each account has different workspaces (ideally different number of them).
User scenario:

0. Logout of all oc and kubectl sessions. After each of the steps perform `chectl workspace:list` and `chectl context:list` commands
1. Login into `s1u2`: `chectl auth:login <s1> -t <refresh-token>`.
2. Login into `s2u1`: `chectl auth:login <s2> -u <username> -p <password>`
3. Login into `s2u2`: `chectl auth:login <s2> -u <username>` (password should be asked)
4. Login by `oc` into `s1u1`. Run `chectl auth:login` (if u1 is not admin user, provide `<s1>` argument)
5. Switch context to another user. For example: `chectl auth:use <s2> -u <u2>`
6. Logout from non current instance by `chectl auth:delete <s1> -u <u1>`
7. Logout from current session by `chectl auth:logout`
8. Try to login into `s3` which is a single user Che installation
9. Try to run `chectl workspace:list --che-api-endpoint <s3>` for single user instance.
10. Another tried use cases are welcomed.

See [docs](https://www.eclipse.org/che/docs/che-7/administration-guide/authenticating-users/) how to get refresh token manually and feel the difference =)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
